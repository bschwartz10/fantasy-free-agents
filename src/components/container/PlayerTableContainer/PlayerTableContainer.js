import React, { Component } from 'react';

import { fetchPlayers } from './apiCalls';
import { fetchADP } from './apiCalls';

import PlayerTable from '../../presentational/PlayerTable';
import PlayerTableFilters from '../../presentational/PlayerTableFilters';

import Loading from '../../presentational/Loading'

class PlayerTableContainer extends Component {
  state = {
    players: [],
    allPlayers: [],
    positions: [],
    loading: true,
    adpSorted: true,
  }
  componentDidMount = () => {
    this.fetch()
  }

  fetch = async () => {
    const rawPlayers = await fetchPlayers()
    const rawAdp = await fetchADP()

    const adp = rawAdp.reduce((agg, playerAdp) => {
      agg[playerAdp.id] = playerAdp.averagePick
      return agg
    },{})

    const players = rawPlayers
                      .map((player) => ({
                        ...player,
                        name: `${player.name.split(',')[1].trim()} ${player.name.split(',')[0]}`,
                        adp: adp[player.id] ? adp[player.id] : '400'
                      }))
                      .sort((firstPlayer, secondPlayer) => {
                        return firstPlayer.adp - secondPlayer.adp
                      })

    this.setState({ players: players, allPlayers: players, loading: false });
    console.log(this.state.players);
  }

  handleClick = (e, type, attribute) => {
    this.sortPlayers(type, attribute)
  }

  sortPlayers = (type, attribute) => {
    const sorted = this.state[type]
    const sortedCollection =
      attribute === 'adp' ?
        this.sortByAdp(sorted, attribute) :
        this.sortByAttribute(sorted, attribute)

    this.setState( {players: sortedCollection, [type]: !sorted} )
  }

  sortByAttribute = (sorted, attribute) => {
    const { players } = this.state
    let sortedCollection;
    if (sorted) {
      sortedCollection = players.sort((firstPlayer, secondPlayer) => {
        if(firstPlayer[attribute] <= secondPlayer[attribute]) return 1
        if(firstPlayer[attribute] >= secondPlayer[attribute]) return -1
      })
    } else {
        sortedCollection = players.sort((firstPlayer, secondPlayer) => {
          if(firstPlayer[attribute] <= secondPlayer[attribute]) return -1
          if(firstPlayer[attribute] >= secondPlayer[attribute]) return 1
        })
    }
    return sortedCollection
  }

  sortByAdp = (sorted, attribute) => {
    const { players } = this.state
    let sortedCollection;
    if (sorted) {
      sortedCollection = players.sort((firstPlayer, secondPlayer) => {
        return secondPlayer[attribute] - firstPlayer[attribute]
      })
    } else {
        sortedCollection = players.sort((firstPlayer, secondPlayer) => {
          return firstPlayer[attribute] - secondPlayer[attribute]
        })
    }
    return sortedCollection
  }

  toggleChange = (e, position) => {
    let { positions } = this.state
    const index = positions.indexOf(position)
    if (index > -1) {
      positions.splice(index, 1)
    } else {
        positions = [...this.state.positions, position]
      }
    this.setState({ positions }, () => {
      this.filterPlayersByPosition()
    })
  }

  filterPlayersByPosition = () => {
    let players = this.state.allPlayers
    let { positions } = this.state
    if (positions.length >= 1) {
      players = players.filter((player) => positions.includes(player.position))
      this.setState({ players })
    } else {
        this.setState({ players: this.state.allPlayers })
    }
  }

  handleChange = (e) => {
    const { value } = e.target
    let allPlayers = this.state.allPlayers
    const updatedList = allPlayers.filter((player) => {
      return player.name.toLowerCase().search(value.toLowerCase()) !== -1
    })
    this.setState({players: updatedList})
  }

  render() {
    const { players, loading } = this.state;
    if (loading) return <Loading />
    return (
      <div>
        <PlayerTableFilters toggleChange={this.toggleChange} handleChange={this.handleChange} />
        <PlayerTable players={players} handleClick={this.handleClick} />
      </div>
    );
  }
}

export default PlayerTableContainer;
