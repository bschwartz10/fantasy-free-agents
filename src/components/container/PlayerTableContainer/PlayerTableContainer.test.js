import React from 'react'
import { shallow } from 'enzyme'
import PlayerTableContainer from './PlayerTableContainer'

describe('PlayerTableContainer component', () => {
  let wrapper;
  let mockPlayer;
  let mockPlayers;

  beforeEach(() => {
    wrapper = shallow(<PlayerTableContainer />)
    wrapper.setState({players: mockPlayers, allPlayers: mockPlayers, loading: false})
    mockPlayer = {
      adp: "400",
      birthdate: "656226000",
      cbs_id: "1977109",
      college: "UCLA",
      draft_team: "FA",
      draft_year: "2012",
      espn_id: "15351",
      height: "72",
      id: "11524",
      jersey: "48",
      kffl_id: "26333",
      name: "Derrick Coleman",
      nfl_id: "derrickcoleman/2534871",
      position: "RB",
      rotoworld_id: "8019",
      sportsdata_id: "e0a63251-5428-43a1-88c1-c000215ac5ce",
      stats_global_id: "461705",
      stats_id: "26299",
      team: "ARI",
      weight: "233",
    }
    mockPlayers = [
      {
        adp: "400",
        birthdate: "656226000",
        cbs_id: "1977109",
        college: "UCLA",
        draft_team: "FA",
        draft_year: "2012",
        espn_id: "15351",
        height: "72",
        id: "11524",
        jersey: "48",
        kffl_id: "26333",
        name: "Derrick Coleman",
        nfl_id: "derrickcoleman/2534871",
        position: "RB",
        rotoworld_id: "8019",
        sportsdata_id: "e0a63251-5428-43a1-88c1-c000215ac5ce",
        stats_global_id: "461705",
        stats_id: "26299",
        team: "ARI",
        weight: "233"
      },
      {
        adp: "1.38",
        birthdate: "775890000",
        cbs_id: "2000877",
        college: "Georgia",
        draft_team: "STL",
        draft_year: "2015",
        espn_id: "2977644",
        height: "73",
        id: "12150",
        jersey: "30",
        kffl_id: "32621",
        name: "Todd Gurley",
        position: "RB",
        rotoworld_id: "10288",
        sportsdata_id: "14263792-d1d3-4b0c-85f7-2a85b4aed6f1",
        stats_global_id: "694641",
        stats_id: "28398",
        team: "LAR",
        weight: "227",
      },
      {
        adp: "118.48",
        birthdate: "658990800",
        cbs_id: "1691489",
        college: "Texas",
        draft_team: "BUF",
        draft_year: "2013",
        espn_id: "15839",
        height: "69",
        id: "11239",
        jersey: "11",
        kffl_id: "27802",
        name: "Marquise Goodwin",
        position: "WR",
        rotoworld_id: "8461",
        sportsdata_id: "bf52ff53-35a6-4696-ac6d-3fa952dc2c87",
        stats_global_id: "507478",
        stats_id: "26701",
        team: "SFO",
        weight: "179",
      }
    ]
  })

  it('should exist', () => {
    expect(wrapper).toBeDefined()
  })

  it('should match snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('sorting players by adp will list players in descending order', () => {
    wrapper.instance().sortPlayers('adpSorted', 'adp')
    expect(wrapper.state('players')[0].name).toEqual('Derrick Coleman')
  })

  it('sorting players by weight will list players in ascending order', () => {
    wrapper.instance().sortPlayers('weightSorted', 'weight')
    expect(wrapper.state('players')[0].name).toEqual('Marquise Goodwin')
  })

  it('filters players by one position', () => {
    wrapper.setState({positions: ['RB']})
    wrapper.instance().filterPlayersByPosition()
    expect(wrapper.state('players').length).toEqual(2)
  })

  it('filters players by two positions', () => {
    wrapper.setState({positions: ['RB', 'WR']})
    wrapper.instance().filterPlayersByPosition()
    expect(wrapper.state('players').length).toEqual(3)
  })

  it('filters players by substring of name', () => {
    const e = {
      target: {
        value: 'Good'
      }
    }
    wrapper.instance().handleChange(e)
    expect(wrapper.state('players').length).toEqual(1)
  })

})
