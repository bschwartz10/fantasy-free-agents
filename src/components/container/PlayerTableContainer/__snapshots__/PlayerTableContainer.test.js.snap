// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerTableContainer component should match snapshot with all data passed in correctly 1`] = `
<div>
  <PlayerTableFilters
    handleChange={[Function]}
    toggleChange={[Function]}
  />
  <PlayerTable
    handleClick={[Function]}
    players={
      Array [
        Object {
          "adp": "400",
          "birthdate": "656226000",
          "cbs_id": "1977109",
          "college": "UCLA",
          "draft_team": "FA",
          "draft_year": "2012",
          "espn_id": "15351",
          "height": "72",
          "id": "11524",
          "jersey": "48",
          "kffl_id": "26333",
          "name": "Derrick Coleman",
          "nfl_id": "derrickcoleman/2534871",
          "position": "RB",
          "rotoworld_id": "8019",
          "sportsdata_id": "e0a63251-5428-43a1-88c1-c000215ac5ce",
          "stats_global_id": "461705",
          "stats_id": "26299",
          "team": "ARI",
          "weight": "233",
        },
        Object {
          "adp": "1.38",
          "birthdate": "775890000",
          "cbs_id": "2000877",
          "college": "Georgia",
          "draft_team": "STL",
          "draft_year": "2015",
          "espn_id": "2977644",
          "height": "73",
          "id": "12150",
          "jersey": "30",
          "kffl_id": "32621",
          "name": "Todd Gurley",
          "position": "RB",
          "rotoworld_id": "10288",
          "sportsdata_id": "14263792-d1d3-4b0c-85f7-2a85b4aed6f1",
          "stats_global_id": "694641",
          "stats_id": "28398",
          "team": "LAR",
          "weight": "227",
        },
        Object {
          "adp": "118.48",
          "birthdate": "658990800",
          "cbs_id": "1691489",
          "college": "Texas",
          "draft_team": "BUF",
          "draft_year": "2013",
          "espn_id": "15839",
          "height": "69",
          "id": "11239",
          "jersey": "11",
          "kffl_id": "27802",
          "name": "Marquise Goodwin",
          "position": "WR",
          "rotoworld_id": "8461",
          "sportsdata_id": "bf52ff53-35a6-4696-ac6d-3fa952dc2c87",
          "stats_global_id": "507478",
          "stats_id": "26701",
          "team": "SFO",
          "weight": "179",
        },
      ]
    }
  />
</div>
`;
